### DISCUSSION TITLE
Mocking directive

### DISCUSSION THREAD

**Discussion by grosch** [help-me-mom/ng-mocks#3000] [Q&A] [answered]
Testing with the below results in this error:

> Error: Unexpected directive 'MockOfLinkDirective' imported by the module 'DynamicTestModule'. Please add an @NgModule annotation.

```typescript
import {Component, Directive, Input} from '@angular/core'
import {createComponentFactory, Spectator} from '@ngneat/spectator/jest'
import {MockBuilder} from 'ng-mocks'

@Directive({
    selector: 'a[appLink]',
    standalone: true
})
export class LinkDirective {
    @Input() appLink!: string
}

@Component({
    selector: '',
    standalone: true,
    imports: [LinkDirective],
    template: '<a appLink="foo"></a>'
})
class ExampleComponent {
}

describe('ExampleComponent', () => {
    let spectator: Spectator<ExampleComponent>
    let component: ExampleComponent

    const dependencies = MockBuilder([ExampleComponent], [LinkDirective]).build()

    const createComponent = createComponentFactory({
        component: ExampleComponent,
        ...dependencies
    })

    beforeEach(() => {
        spectator = createComponent()
        component = spectator.component
    })

    it('should create', () => {
        expect(component).toBeTruthy()
    })
})
```

Based on your reply to the other discussion, I thought you might mean I had to do this instead:

```typescript
    const dependencies = MockBuilder(ExampleComponent)
        .mock(LinkDirective)
        .build()
```

but that has the same error.

**Comment by satanTime** [comment, marked as answer]
## Updated

I think, it has been fixed already by https://github.com/help-me-mom/ng-mocks/releases/tag/v14.1.0, could you verify if it works for you?

## Original

In your example, because `LinkDirective` has been imported by `ExampleComponent`, and you want to mock it, you should simply use:

```ts
const dependencies = MockBuilder(ExampleComponent).build();
```

I know that docs are hard to read, I'm very sorry for that, simply not much time to improve them.
More info why and how is here: https://ng-mocks.sudo.eu/guides/component-standalone

Anyway, `MockBuilder([ExampleComponent], [LinkDirective]).build()` shouldn't throw, I'll double check what is wrong there and I'll try to fix it.

**Reply by grosch** [reply]
Your docs are actually quite good. The concepts are just new-ish to me, so I struggle to know exactly where to look and then understand what it's saying.  So if I follow correctly, my standalone component imports everything it needs to work properly.  Now, when I MockBuilder that component, you automatically mock everything that it imports.  

However, sometimes that doesn't work.  For example, I import IgxInputGroupModule and need to do testing against it, so I'd add that to the first parameter's array item, and then you would not mock that.  Am I following?

```typescript
MockBuilder([MyStandaloneComponent, IgxInputGroupModule])
```

What's the difference between listing it like I did above, and using `.keep(IgxInputGroupModule)` like you show in the link, above?

The second argument is the array of items to mock. If you've already explicitly mocked everything, when would I use that second parameter? It seems like I might use that with non-standalone components, but if it's standalone, I'd never use it.

**Reply by grosch** [reply]
BTW, using the line you suggested:

```typescript
const dependencies = MockBuilder(ExampleComponent).build();
```

> Error: The "MockOfLinkDirective" directive, imported from "ExampleComponent", is not standalone. Did you forget to add the standalone: true flag?

**Reply by satanTime** [reply]
Thanks for checking. I'm flying this week. On the weekend, I should have more info to share with you. 

It looks like there is bug with internal mocking.

**Reply by grosch** [reply]
Any progress on this one?

**Reply by satanTime** [reply]
I think, it has been fixed already by https://github.com/help-me-mom/ng-mocks/releases/tag/v14.1.0, could you verify if it works for you?

**Reply by grosch** [reply]
Yup, seems to be working, thanks!
